### PR TITLE
[FEATURE] Make constraint reporting less obtrusive

### PR DESCRIPTION
**Pull request by m-kuhn**
A new and polished look and feel for the constraints. They are now only shown inline and the top widget showing up to three constraints has been removed. The constraint text is reported as a tooltip.

Do we need a list of all constraints still somewhere?
We could add a button at the bottom "constraint report" that would open up such a dialog, but I'm not sure that's actually required.

unobtrusiveconstraints

### CODE REVIEW

File: tests/src/gui/testqgsattributeform.cpp [hunk 1]
```diff
@@ -84,9 +92,9 @@ void TestQgsAttributeForm::testFieldConstraint()
   form.setFeature( ft );
 
   // testing stuff
-  QString validLabel = QStringLiteral( "col0<font color=\"green\">✔</font>" );
-  QString invalidLabel = QStringLiteral( "col0<font color=\"red\">✘</font>" );
-  QString warningLabel = QStringLiteral( "col0<font color=\"orange\">✘</font>" );
+  QString validLabel = QStringLiteral( "<font color=\"#259b24\">\u2714</font>" );
+  QString invalidLabel = QStringLiteral( "<font color=\"#FF9800\">\u2718</font>" );
```
Comment (nyalldawson) [inline tests/src/gui/testqgsattributeform.cpp]: Why orange here, not red? I think this should be something like b00000

File: src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp [hunk 2]
```diff
@@ -110,11 +110,11 @@ void QgsEditorWidgetWrapper::updateConstraintWidgetStatus( ConstraintResult cons
       break;
 
     case ConstraintResultFailHard:
-      widget()->setStyleSheet( QStringLiteral( "background-color: #dd7777;" ) );
+      widget()->setStyleSheet( QStringLiteral( "background-color: #FFE0B2;" ) );
```
Comment (nyalldawson) [inline src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp]: Again, why the switch from red to yellow?

File: src/gui/qgsattributeformeditorwidget.cpp [hunk 3]
```diff
@@ -37,6 +37,10 @@ QgsAttributeFormEditorWidget::QgsAttributeFormEditorWidget( QgsEditorWidgetWrapp
   , mIsMixed( false )
   , mIsChanged( false )
 {
+  mConstraintResultLabel = new QLabel();
```
Comment (nyalldawson) [inline src/gui/qgsattributeformeditorwidget.cpp]: This looks leaky - as far as i can tell, the parent is not set until `setMode( MultiEditMode )` is called (which may be never).

File: src/gui/qgsattributeformeditorwidget.cpp [hunk 4]
```diff
@@ -134,6 +138,27 @@ QList< QgsSearchWidgetWrapper * > QgsAttributeFormEditorWidget::searchWidgetWrap
   return mSearchWidgets;
 }
 
+void QgsAttributeFormEditorWidget::setConstraintStatus( const QString &constraint, const QString &description, const QString &err, QgsEditorWidgetWrapper::ConstraintResult result )
+{
+  switch ( result )
+  {
+    case QgsEditorWidgetWrapper::ConstraintResultFailHard:
+      mConstraintResultLabel->setText( QStringLiteral( "<font color=\"#FF9800\">✘</font>" ) );
```
Comment (nyalldawson) [inline src/gui/qgsattributeformeditorwidget.cpp]: red?

File: src/gui/qgsattributeformeditorwidget.h [hunk 5]
```diff
@@ -202,6 +209,7 @@ class GUI_EXPORT QgsAttributeFormEditorWidget : public QWidget
     QgsEditorWidgetWrapper *mWidget = nullptr;
     QList< QgsSearchWidgetWrapper * > mSearchWidgets;
     QgsAttributeForm *mForm = nullptr;
+    QLabel *mConstraintResultLabel;
```
Comment (nyalldawson) [inline src/gui/qgsattributeformeditorwidget.h]: = nullptr

**Comment by nyalldawson** [comment]
Looks MUCH improved to me, nice! I'm just unsure about the logic behind the change from red for a failing hard constraint to yellow.

**Comment by nyalldawson** [comment]
> Do we need a list of all constraints still somewhere?
We could add a button at the bottom "constraint report" that would open up such a dialog, but I'm not sure that's actually required.

I agree

**Comment by m-kuhn** [comment]
Red indicates "error", a condition from which it is hard to recover and where something has gone wrong.
What we want to do here is to invite the user to kindly fill in missing information.

**Comment by nyalldawson** [comment]
Isn't this an error? I'd say it is. It's preventing the user from adding/editing the feature... The yellow looks to me more like a friendly warning, but this is more severe than that.

**Comment by m-kuhn** [comment]
But we dob't want to scare him. Every single time he adfs a new feature. Digitizing should be fun. All those required password bozes on signup pages also only become red once you clicked the submit button and validation failed.

**Comment by pblottiere** [comment]
> It's preventing the user from adding/editing the feature... The yellow looks to me more like a friendly warning, but this is more severe than that.

My 2 cents : I agree pretty much with Nyall.

**Comment by nirvn** [comment]
I'd also go for a red color. It's an established visual queue (HTML web forms, etc.) that is worth sticking to. However, a strong #FF0000 can be toned done a bit: http://www.colourlovers.com/colors/search?query=red 😄

**Comment by pblottiere** [comment]
> They are now only shown inline and the top widget showing up to three constraints has been removed. The constraint text is reported as a tooltip.

And by the way, it's much nicer this way!

**Comment by m-kuhn** [comment]
The reason for the change is, that I got a user request, about being annoyed by the red, just to make the point that I'm not defending my lonely warrior standpoint here ;)

To repeat, all good forms on the web I know will start with no red (or just a hint (like a char) next to the field. Once you click ok, they go red but not before. (see below)

I hate software that starts in an erroneous state. To me as a user red means "you have done something wrong. Try doing better next time.". So, if any change, I'd propose to change the color upon Ok click.

github signup
signup google

**Comment by pblottiere** [comment]
Dumb idea : maybe we could have a different behavior in case of a feature edition and a feature addition? In case of a new feature, it's true that it may be annoying to have 'too much red'. But in case of a feature edition, user could/should be immediately notified in case of an invalid field?

**Comment by m-kuhn** [comment]
> In case of a new feature, it's true that it may be annoying to have 'too much red'. But in case of a feature edition, user could/should be immediately notified in case of an invalid field?

They are notified anyway. Immediately.

> Dumb idea : maybe we could have a different behavior in case of a feature edition and a feature addition? 

If

1. red is really a requirement
2. everyone agrees that red shouldn't be shown on opening a new "add feature" form
3. we go the road of having some (unobtrusive) hints on an empty new feature form
4. and subsequently changing it to red only once the user tried to click the (always enabled) ok button

then yes :)

**Comment by m-kuhn** [comment]
Ok, folks.

I'll go ahead merge this because it seems functionality-wise it's unproblematic and only colors need to be discussed, which can also be done during freeze.
Please give it a try and make your mind how it feels in real life. As well as leave your comments on the above idea.

**Pull request merged**